apisix/stream/xrpc/protocols/redis/init.lua

```diff
@@ -0,0 +1,239 @@
+        return nil, err
+    end
+
+    local ctx = {
+        cmd_line = core.table.new(narg, 0)
+    }
+
+    for i = 1, narg do
+        local n, err = read_len(sk)
+        if not n then
+            return nil, err
+        end
+
+        local p, err = sk:read(n + 2)
+        if not p then
+            return nil, err
+        end
+
+        local s = ffi_str(p, n)
+        ctx.cmd_line[i] = s
+    end
+
+    ctx.cmd = ctx.cmd_line[1]
+    return ctx
+end
+
+
+local function read_reply(sk)
+    local line, err, n = read_line(sk)
+    if not line then
+        return nil, err
+    end
+
+    local prefix = line[0]
+
+    if prefix == PREFIX_STR then    -- char '$'
+        -- print("bulk reply")
+
+        local size = tonumber(ffi_str(line + 1, n - 1))
+        if size < 0 then
+            -- return null
+            return true
+        end
+
+        local ok, err = sk:drain(size + 2)
+        if not ok then
+            return nil, err
+        end
+
+        return true
+
+    elseif prefix == PREFIX_STA then    -- char '+'
+        -- print("status reply")
+        -- return sub(line, 2)
+        return true
+
+    elseif prefix == PREFIX_ARR then -- char '*'
+        local narr = tonumber(ffi_str(line + 1, n - 1))
+
+        -- print("multi-bulk reply: ", narr)
+        if narr < 0 then
+            -- return null
+            return true
+        end
+
+        local vals = core.table.new(n, 0)
+        local nvals = 0
+        for i = 1, narr do
+            local res, err = read_reply(sk)
+            if res then
+                nvals = nvals + 1
+                vals[nvals] = res
+
+            elseif res == nil then
+                return nil, err
+
+            else
+                -- be a valid redis error value
+                nvals = nvals + 1
+                vals[nvals] = {false, err}
+            end
+        end
+
+        return vals
+
+    elseif prefix == PREFIX_INT then    -- char ':'
+        -- print("integer reply")
+        -- return tonumber(str_sub(line, 2))
+        return true
+
+    elseif prefix == PREFIX_ERR then    -- char '-'
+        -- print("error reply: ", n)
+        -- return false, str_sub(line, 2)
+        return true
+
+    else
+        return nil, str_fmt("unknown prefix: \"%s\"", prefix)
+    end
+end
+
+
+function _M.from_downstream(session, downstream)
+    local ctx, err = read_req(downstream)
+    if not ctx then
+        if err ~= "timeout" and err ~= "closed" then
+            core.log.error("failed to read request: ", err)
+        end
+        return DECLINED
+    end
+
+    return OK, ctx
+end
+
+
+function _M.connect_upstream(session, ctx)
+    local conf = session.upstream_conf
+    local nodes = conf.nodes
+    if #nodes == 0 then
+        core.log.error("failed to connect: no nodes")
+        return DECLINED
+    end
+
+    local node = nodes[math_random(#nodes)]
+    local sk = sdk.connect_upstream(node, conf)
+    if not sk then
+        return DECLINED
+    end
```

Comment:
Would be better to have a retry logic here ? 

A detailed error message is also required. The current one is not clear which node was accessed at the moment.